File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Point implements Comparable<Point>{
    int x;
    int y;
    int count; // 움직인 횟수
    Point(int x, int y, int count) {
        this.x = x;
        this.y = y;
        this.count = count;
    }
    @Override
    public int compareTo(Point p) {
        return Integer.compare(this.count, p.count);
    }
}

class Solution {
    public int n, m;
    public boolean visited[][];
    public int dx[] = {-1,1,0,0}; // 방향벡터
    public int dy[] = {0,0,-1,1};
    
    public int solution(int[][] maps) {
        int answer = -1;
        n = maps.length;
        m = maps[0].length;
        
        visited = new boolean[n][m]; // 방문 배열
        PriorityQueue<Point> queue = new PriorityQueue<>();
        queue.offer(new Point(0,0,1));
        visited[0][0] = true;
        
        while(!queue.isEmpty()) {
            Point q = queue.poll();
            
            if(q.x==n-1 && q.y==m-1) {
                answer = q.count;
                break;
            }
            
            for(int i=0; i<4; i++) {
                int nx = q.x + dx[i];
                int ny = q.y + dy[i];

                if((nx>=0&&nx<n)&&(ny>=0&&ny<m)&&!visited[nx][ny]&&maps[nx][ny]==1) {
                    queue.offer(new Point(nx, ny, q.count+1));
                    visited[nx][ny] = true;
                }
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let answer = 0
    let n = maps.length
    let m = maps[0].length
    
    // (n-1,m-1) 도착
    // 방문 배열
    let visited = new Array(n).fill(false).map(v => new Array(m).fill(false))
    let direction = [{x:1,y:0}, {x:-1,y:0}, {x:0,y:1}, {x:0,y:-1}]
    
    // 큐
    let queue = []
    // 출발지
    queue.push({x:0, y:0, move: 1})
    visited[0][0] = true
    
    while(queue.length > 0) {
        let q = queue.shift()

        if(q.x === n-1 && q.y === m-1) return q.move
        
        direction.map((v) => {
            let nx = q.x + v.x
            let ny = q.y + v.y

            if(nx>=0&&nx<=n-1&&ny>=0&&ny<=m-1&&!visited[nx][ny]&&maps[nx][ny]===1) {
                visited[nx][ny] = true
                queue.push({x:nx, y:ny, move: q.move + 1})
            }
        })
    }
    
    return -1;
}